### PR TITLE
Port `arrow_typeof` to datafusion-function

### PR DESCRIPTION
**Pull request by yyy1000**
## Which issue does this PR close?

Related #9285.

### CODE REVIEW

**Comment by yyy1000** [comment]
The CI fails because https://github.com/apache/arrow-datafusion/pull/9388/files#r1506662726,
I'm wondering whether I should change the test case or wait for #9392 is implemented. 🤔

**Comment by yyy1000** [comment]
> The CI fails because https://github.com/apache/arrow-datafusion/pull/9388/files#r1506662726, I'm wondering whether I should change the test case or wait for #9392 is implemented. 🤔

Update: I think changing the test case now and revise it later is better. Because now it corresponds to the expected behavior since hint for UDF is not supported. And after hint is supported, we will revise more test cases besides this.

File: datafusion/functions/src/core/nullif.rs [hunk 1]
```diff
@@ -15,8 +15,6 @@
 // specific language governing permissions and limitations
 // under the License.
 
-//! Encoding expressions
```
Comment (alamb) [inline datafusion/functions/src/core/nullif.rs:18]: 👍

File: datafusion/sqllogictest/test_files/functions.slt [hunk 2]
```diff
@@ -479,7 +479,7 @@ CREATE TABLE test(
 (3, 30);
 
 # Scalar function
-statement error Did you mean 'arrow_typeof'?
+statement error Invalid function 'arrowtypeof'
```
Comment (alamb) [inline datafusion/sqllogictest/test_files/functions.slt:482]: I think this is tracked by https://github.com/apache/arrow-datafusion/issues/9392

@SteveLauC  has a PR up for it https://github.com/apache/arrow-datafusion/pull/9407 which hopefully will land shortly

File: docs/source/user-guide/sql/data_types.md [hunk 3]
```diff
@@ -31,7 +31,7 @@ the `arrow_typeof` function. For example:
 ```sql
 select arrow_typeof(interval '1 month');
 +-------------------------------------+
-| arrowtypeof(IntervalYearMonth("1")) |
+| arrow_typeof(IntervalYearMonth("1")) |
```
Comment (alamb) [inline docs/source/user-guide/sql/data_types.md:34]: 👍

**Review by alamb — APPROVED** [review]
Thank you very much @yyy1000  -- this looks good to me

File: datafusion/sqllogictest/test_files/functions.slt (hunk 2 above)
Comment (alamb) [reply datafusion/sqllogictest/test_files/functions.slt:482]: https://github.com/apache/arrow-datafusion/pull/9407 is ready to go -- I think once we merge that one we can revert this change

Comment (alamb) [reply datafusion/sqllogictest/test_files/functions.slt]: 3dbafa70e

**Comment by alamb** [comment]
I took the liberty of merging this PR up from main and resolving conflicts (also I haven't been coding much this week and I wanted some excuse to do so)

**Comment by alamb** [comment]
The Ci failure seems unrelated https://github.com/apache/arrow-datafusion/actions/runs/8221664632/job/22482282908?pr=9524

I will rerun the windows test

**Comment by jayzhan211** [comment]
> The Ci failure seems unrelated https://github.com/apache/arrow-datafusion/actions/runs/8221664632/job/22482282908?pr=9524
> 
> I will rerun the windows test

I had meet this error several times, but it is not reproducible. It only appears at the first time after compilation, but not always 😕

**Comment by alamb** [comment]
> > The Ci failure seems unrelated https://github.com/apache/arrow-datafusion/actions/runs/8221664632/job/22482282908?pr=9524
> > I will rerun the windows test
> 
> I had meet this error several times, but it is not reproducible. It only appears at the first time after compilation, but not always 😕

I think it is likely a intermittent failure related to the test output not being deterministic (like maybe it depends on the order of thread execution or something). It is probably worth a ticket to track it (but I am about out of time today)

**Comment by yyy1000** [comment]
> I took the liberty of merging this PR up from main and resolving conflicts (also I haven't been coding much this week and I wanted some excuse to do so)

Thanks! @alamb

**Comment by yyy1000** [comment]
I resolved the conflict with main branch :)

**Comment by alamb** [comment]
> I resolved the conflict with main branch :)

We are sooooo close 🙏

**Comment by alamb** [comment]
Thanks again @yyy1000

**Pull request merged**